glotaran/analysis/util.py

```diff
@@ -160,55 +179,78 @@ def retrieve_clps(
+            clps[target_idx] = relation.parameter * clps[source_idx]
     return clps
 
 
 def calculate_clp_penalties(
     model: Model,
     parameters: ParameterGroup,
-    clps: xr.DataArray,
-    global_dimension: str,
+    clp_labels: list[list[str]] | list[str],
+    clps: list[np.ndarray],
+    global_axis: np.ndarray,
 ) -> np.ndarray:
 
     penalties = []
     for penalty in model.clp_area_penalties:
-        if (
-            penalty.source in clps.coords["clp_label"]
-            and penalty.target in clps.coords["clp_label"]
-        ):
-            penalty = penalty.fill(model, parameters)
-
-            source_area = xr.concat(
-                [
-                    clps.sel(
-                        {
-                            "clp_label": penalty.source,
-                            global_dimension: slice(interval[0], interval[1]),
-                        }
-                    )
-                    for interval in penalty.source_intervals
-                ],
-                dim=global_dimension,
-            )
-
-            target_area = xr.concat(
-                [
-                    clps.sel(
-                        {
-                            "clp_label": penalty.target,
-                            global_dimension: slice(interval[0], interval[1]),
-                        }
-                    )
-                    for interval in penalty.target_intervals
-                ],
-                dim=global_dimension,
-            )
-
-            area_penalty = np.abs(np.sum(source_area) - penalty.parameter * np.sum(target_area))
-            penalties.append(area_penalty * penalty.weight)
+        penalty = penalty.fill(model, parameters)
+        source_area = _get_area(
+            penalty.source,
+            clp_labels,
+            clps,
+            penalty.source_intervals,
+            global_axis,
+        )
+
+        target_area = _get_area(
+            penalty.target,
+            clp_labels,
+            clps,
+            penalty.target_intervals,
+            global_axis,
+        )
+
+        area_penalty = np.abs(np.sum(source_area) - penalty.parameter * np.sum(target_area))
+
+        penalties.append(area_penalty * penalty.weight)
 
     return np.asarray(penalties)
+
+
+def _get_area(
+    clp_label: str,
+    clp_labels: list[list[str]],
+    clps: list[np.ndarray],
+    intervals: list[tuple[float, float]],
+    global_axis: np.ndarray,
+) -> np.ndarray:
+    area = []
+
+    for interval in intervals:
+        if interval[0] > global_axis[-1]:
+            continue
+
+        start_idx, end_idx = get_idx_from_interval(interval, global_axis)
+        for i in range(start_idx, end_idx + 1):
+            index_clp_labels = clp_labels[i] if isinstance(clp_labels[0], list) else clp_labels
+            if clp_label in index_clp_labels:
+                area.append(clps[i][index_clp_labels.index(clp_label)])
+
+    return np.asarray(area)  # TODO: normalize for distance on global axis
```

Comment:
```suggestion
    return np.sum(area)  # TODO: normalize for distance on global axis
```
`area` being a float makes more sense and we would need to call `np.sum` on it in `calculate_clp_penalties`.
Same is true for `_get_area` in `glotaran/model/clp_penalties.py`.

Reply:
Nope, it is really needed for the the optimization. Each penalty is not just added to the sum, but rather increases the size of the penalty vector.

Reply:
Ok, after personal discussion with @s-weigand we decided wo make a decicated issue for that.

Reply:
https://github.com/glotaran/pyglotaran/issues/746